File: backend/app/schemas/route.py
from pydantic import BaseModel, Field
from typing import List, Optional, Tuple, Dict, Any
from datetime import datetime

class RoutePreferences(BaseModel):
    prioritize: str = "balanced"  # "safety", "speed", "balanced"
    avoid_police: bool = False
    max_detour_minutes: int = 10
    show_police_checkpoints: bool = False

class RouteAnalysisRequest(BaseModel):
    origin: Tuple[float, float] = Field(..., description="(longitude, latitude)")
    destination: Tuple[float, float] = Field(..., description="(longitude, latitude)")
    ride_type: str = "moto"  # "moto", "economy_car", "comfort_car"
    departure_time: Optional[datetime] = None
    preferences: RoutePreferences = RoutePreferences()

class RouteIncident(BaseModel):
    id: str
    type: str
    severity: str  # "low", "moderate", "severe"
    confirmations: int
    distance_from_start_km: float
    estimated_delay_minutes: int
    location: Tuple[float, float]
    description: Optional[str] = None

class ScoredRoute(BaseModel):
    id: str
    geometry_encoded: str  # Polyline string
    distance_km: float
    duration_minutes: int
    score: int  # 0-100
    rank: int
    label: str
    incidents: List[RouteIncident]
    warnings: List[str]
    historical_notes: List[str]

class RouteAnalysisResponse(BaseModel):
    routes: List[ScoredRoute]
    recommendation: Optional[str] = None

File: mobile/src/services/route.ts
import { api } from './api';

export interface RouteIncident {
    id: string;
    type: string;
    severity: string;
    description?: string;
    location: [number, number];
}

export interface ScoredRoute {
    id: string;
    geometry_encoded: string;
    distance_km: number;
    duration_minutes: number;
    score: number;
    rank: number;
    label: string;
    incidents: RouteIncident[];
    warnings: string[];
}

export interface RouteAnalysisResponse {
    routes: ScoredRoute[];
    recommendation?: string;
}

export const routeService = {
    analyzeRoutes: async (
        origin: { latitude: number; longitude: number }, 
        destination: { latitude: number; longitude: number },
        rideType: 'moto' | 'economy_car' | 'comfort_car' = 'moto'
    ): Promise<RouteAnalysisResponse> => {
        try {
            const response = await api.post('/routes/analyze', {
                origin: [origin.longitude, origin.latitude],
                destination: [destination.longitude, destination.latitude],
                ride_type: rideType,
                preferences: {
                    prioritize: 'balanced'
                }
            });
            return response.data;
        } catch (error) {
            console.error('Route analysis failed:', error);
            throw error;
        }
    }
};
